core/src/main/java/eu/maveniverse/maven/njord/shared/impl/DefaultSession.java

```diff
@@ -124,9 +129,39 @@ public Optional<Model> readEffectiveModel(Artifact artifact, List<RemoteReposito
         requireNonNull(artifact);
         requireNonNull(remoteRepositories);
         checkClosed();
+
+        TransferListener transferListener = new AbstractTransferListener() {
+            @Override
+            public void transferSucceeded(TransferEvent event) {
+                String action = (event.getRequestType() == TransferEvent.RequestType.PUT ? "Uploaded" : "Downloaded");
+                String direction = event.getRequestType() == TransferEvent.RequestType.PUT ? "to" : "from";
+                TransferResource resource = event.getResource();
+                logger.debug(
+                        "{} {} {}: {}{}",
+                        action,
+                        direction,
+                        resource.getRepositoryId(),
+                        resource.getRepositoryUrl(),
+                        resource.getResourceName());
+            }
+
+            @Override
+            public void transferFailed(TransferEvent event) {
+                String action = (event.getRequestType() == TransferEvent.RequestType.PUT ? "upload" : "download");
+                String direction = event.getRequestType() == TransferEvent.RequestType.PUT ? "to" : "from";
+                TransferResource resource = event.getResource();
+                logger.warn(
+                        "Failed {} {} {}: {}{}",
+                        action,
+                        direction,
+                        resource.getRepositoryId(),
+                        resource.getRepositoryUrl(),
+                        resource.getResourceName());
+            }
```

Comment:
Also is the error supposed to be triggered here or these listeners are just used for logging?

Reply:
just logging, the error happens usually in transporter (transport issue like HTTP 404, 500, etc) or in basic connector (checksum mismatch + policy that makes it fail).